### ISSUE TITLE
[question] When to specify which `compiler.runtime_version` settings of Clang compilers?

### ISSUE THREAD

**Issue by hwhsu1231** [conan-io/conan#12432]
### Problem Description

According to this [article](https://dev.to/yumetodo/list-of-mscver-and-mscfullver-8nd), there are 4 types of version with MSVC compilers. And the corresponding table is like the following. It seems that `compiler.runtime_version` is related to **MSVC toolset version**. I was wondering what value of `compiler.runtime_version` should I specify for Clang compilers?

| VC version | `_MSC_VER` macro | Visual Studio | MSVC toolset version |
| ---------- | ---------------- | ------------- | -------------------- |
| 8.0  | 1400 | 2005 | 80  |
| 10.0 | 1500 | 2008 | 90  |
| 11.0 | 1600 | 2010 | 100 |
| 12.0 | 1700 | 2012 | 110 |
| 13.0 | 1800 | 2013 | 120 |
| 14.0 | 1900 | 2015 | 140 |
| 15.0 | 1910~1919 | 2017 | 141 |
| 16.0 | 1920~1929 | 2019 | 142 |
| 17.0 | 1930~1939 | 2022 | 143 |

![image](https://user-images.githubusercontent.com/71438617/199062022-63fa0e33-cd11-4200-9eac-df61c3693729.png)

**Comment by hwhsu1231** [comment]
Similar to https://github.com/conan-io/conan/issues/12387, I also want to know what is a suitable detection mechanism to specify `compiler.runtime_version` in [cmake-conan](https://github.com/conan-io/cmake-conan). 

There is a related CMake variable called [MSVC_TOOLSET_VERSION](https://cmake.org/cmake/help/latest/variable/MSVC_TOOLSET_VERSION.html). However, it seems that this variable is set only when we use MSVC compilers. And it is **EMPTY** when I use the following MSVC-based Clang compilers:

```
C:\Program Files\LLVM\bin\clang++.exe
C:\Program Files\LLVM\bin\clang-cl.exe
C:\Program Files (x86)\LLVM\bin\clang++.exe
C:\Program Files (x86)\LLVM\bin\clang-cl.exe
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\Llvm\bin\clang++.exe
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\Llvm\bin\clang-cl.exe
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\Llvm\x64\bin\clang++.exe
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\Llvm\x64\bin\clang-cl.exe
```

**Comment by memsharded** [comment]
> compiler.runtime_version

yes, that is the version of the toolset, so ``v141`` is the toolset version that comes with VS 15.0


> Similar to https://github.com/conan-io/conan/issues/12387, I also want to know what is a suitable detection mechanism to specify compiler.runtime_version in [cmake-conan](https://github.com/conan-io/cmake-conan).

I have no idea, I don't even know if this is possible. This is one of the reasons why I don't like very much the ``cmake-conan`` integration, and trying to guess the setup and settings from the current execution cmake.

**Comment by hwhsu1231** [comment]
@memsharded 

I found that the [MSVC_VERSION](https://cmake.org/cmake/help/latest/variable/MSVC_VERSION.html) variable is still defined even though we use those MSVC-based Clang compilers. Maybe we can use this variable to specify the corresponding `compiler.runtime_version` in cmake-conan. For example, design an internal function called `conan_cmake_detect_clang_runtime_version`. What do you think?

```cmake
function(conan_detect_clang_runtime_version result)
    set(${result} "" PARENT_SCOPE)
    if(NOT MSVC_VERSION VERSION_LESS 1900 AND MSVC_VERSION VERSION_LESS 1910)
        set(${result} "v140" PARENT_SCOPE)
    elseif(NOT MSVC_VERSION VERSION_LESS 1910 AND MSVC_VERSION VERSION_LESS 1920)
        set(${result} "v141" PARENT_SCOPE)
    elseif(NOT MSVC_VERSION VERSION_LESS 1920 AND MSVC_VERSION VERSION_LESS 1930)
        set(${result} "v142" PARENT_SCOPE)
    elseif(NOT MSVC_VERSION VERSION_LESS 1930 AND MSVC_VERSION VERSION_LESS 1940)
        set(${result} "v143" PARENT_SCOPE)
    else()
        message(FATAL_ERROR "Conan: Unknown Clang runtime version [${MSVC_VERSION}]")
    endif()
endfunction()
```

![image](https://user-images.githubusercontent.com/71438617/199140136-b68a219a-abab-49c6-ad92-88b11825cd0c.png)

**Comment by hwhsu1231** [comment]
It seems that [MSVC_TOOLSET_VERSION](https://cmake.org/cmake/help/latest/variable/MSVC_TOOLSET_VERSION.html) is:

- **SPECIFIED** if we use **MSVC-like CLI** Clang of LLVM (`clang-cl.exe`).
- **EMPTY** if we use **GNU-like CLI** Clang of LLVM (`clang++.exe`).

![image](https://user-images.githubusercontent.com/71438617/199159145-99ed34f9-ba2b-4d71-a517-ee76db85d29e.png)

**Comment by memsharded** [comment]
Yes, if MSVC_TOOLSET_VERSION is not defined, then it seems that mapping the MSVC_VERSION would be the only alternative

**Comment by memsharded** [comment]
Following up on this old ticket

Things have changed since then, Conan 2, even the ``cmake-conan`` is a completely new one, based on CMake dependency providers.

Closing this issue as outdated/staled, please open new tickets against Conan 2 (in the ``cmake-conan`` repo if it is related to cmake-conan and not the Conan client) for any further question or issue, thanks for the feedback!
